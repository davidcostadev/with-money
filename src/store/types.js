export const ADD_USER = 'ADD_USER';
export const TO_CURRENT_MONTH = 'TO_CURRENT_MONTH';
export const TO_PREVIUS_MONTH = 'TO_PREVIUS_MONTH';
export const TO_NEXT_MONTH = 'TO_NEXT_MONTH';
