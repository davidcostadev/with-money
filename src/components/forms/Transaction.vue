<template>
  <form>
    <md-card>
      <md-card-content>
        <md-field>
          <label for="name">Name</label>
          <md-input v-model="nameMutable" id="name" />
        </md-field>
        <md-field>
          <input type="text" id="value" v-model="valueMutable" class="md-input" />
          <md-icon>attach_money</md-icon>
        </md-field>
        <md-datepicker v-model="transactionDateMutable" id="transactionDate">
          <label>Transaction Day</label>
        </md-datepicker>
        <md-field>
          <label for="type">Type</label>
          <md-select v-model="typeMutable" name="type" id="type">
            <md-option value="in" id="type-in">In</md-option>
            <md-option value="out" id="type-out">Out</md-option>
          </md-select>
        </md-field>
        <md-field>
          <label for="AccountId">Account</label>
          <md-select v-model="AccountIdMutable" name="AccountId" id="account">
            <md-option
            v-for="account in accounts"
            :key="account.id"
            :value="account.id"
            :id="`account-${account.id}`"
            >
              {{account.name}}
            </md-option>
          </md-select>
        </md-field>
        <md-field>
          <label for="CategoryId">Category</label>
          <md-select v-model="CategoryIdMutable" name="CategoryId" id="category">
            <md-option
            v-for="category in categories"
            :key="category.id"
            :value="category.id"
            :id="`category-${category.id}`"
            >
              {{category.name}}
            </md-option>
          </md-select>
        </md-field>
        <div>
          <md-switch v-model="isPaidMutable" class="md-primary" id="isPaid">IsPaid?</md-switch>
        </div>
      </md-card-content>
      <md-card-actions>
        <md-button
        class="md-primary"
        @click="save"
        :disabled="isLoading"
        id="save"
      >Save</md-button>
      </md-card-actions>
    </md-card>
  </form>
</template>

<script>
import Accounts from '../../services/accounts';
import Categories from '../../services/categories';

export default {
  props: {
    name: {
      type: String,
      default: '',
    },
    value: {
      type: String,
      default: '',
    },
    AccountId: {
      type: Number,
      default: 0,
    },
    CategoryId: {
      type: Number,
      default: 0,
    },
    type: {
      type: String,
      default: 'out',
    },
    isPaid: {
      type: Boolean,
      default: false,
    },
    transactionDate: {
      type: [String, Date],
      default: () => new Date(),
    },
    onSave: {
      type: Function,
      required: true,
    },
  },
  data() {
    return {
      accounts: [],
      categories: [],
      isLoading: false,
      nameMutable: this.name,
      valueMutable: this.value,
      transactionDateMutable: this.transactionDate,
      typeMutable: this.type,
      isPaidMutable: this.isPaid,
      AccountIdMutable: this.AccountId,
      CategoryIdMutable: this.CategoryId,
    };
  },
  methods: {
    async save() {
      this.isLoading = true;

      await this.onSave({
        name: this.nameMutable,
        value: this.valueMutable,
        AccountId: this.AccountIdMutable,
        CategoryId: this.CategoryIdMutable,
        type: this.typeMutable,
        isPaid: this.isPaidMutable,
        transactionDate: this.transactionDateMutable,
      });

      this.isLoading = false;
    },

    async getAccounts() {
      const { data } = await Accounts.list();
      this.accounts = data;
    },
    async getCategories() {
      const { data } = await Categories.list({
        type: this.typeMutable,
      });

      this.categories = data;
    },
  },
  async created() {
    await this.getAccounts();
    await this.getCategories();
  },
};
</script>
