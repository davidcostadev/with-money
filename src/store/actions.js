import * as types from './types';

export default {
  addUser({ commit }, payload) {
    commit(types.ADD_USER, payload);
  },
  toNextMonth({ commit }) {
    commit(types.TO_NEXT_MONTH);
  },
  toCurrentMonth({ commit }) {
    commit(types.TO_CURRENT_MONTH);
  },
  toPreviousMonth({ commit }) {
    commit(types.TO_PREVIUS_MONTH);
  },
};
