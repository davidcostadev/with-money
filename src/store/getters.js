export default {
  user(state) {
    return state.user;
  },
  current_month(state) {
    return state.current_month;
  },
  state_month(state) {
    return state.state_month;
  },
  state_month_str(state) {
    return state.state_month_str;
  },
  next_month_str(state) {
    return state.next_month_str;
  },
  previous_month_str(state) {
    return state.previous_month_str;
  },
};
