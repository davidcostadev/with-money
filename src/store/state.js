import Moment from 'moment';

export default {
  user: null,
  current_month: Moment(),
  state_month: Moment(),
  state_month_str: Moment().format('MMM YY'),
  next_month_str: Moment().add(1, 'month').format('MMM YY'),
  previous_month_str: Moment().subtract(1, 'month').format('MMM YY'),
  flash_message: '',
};
