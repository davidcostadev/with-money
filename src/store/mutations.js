import Moment from 'moment';
import * as types from './types';
import { updateStatesMont } from '../utils/state';

export default {
  [types.ADD_USER](state, payload) {
    state.user = payload;
  },
  [types.TO_CURRENT_MONTH](state) {
    state.state_month = state.current_month;
    updateStatesMont(state);
  },
  [types.TO_PREVIUS_MONTH](state) {
    state.state_month = Moment(state.state_month).subtract(1, 'month');
    updateStatesMont(state);
  },
  [types.TO_NEXT_MONTH](state) {
    state.state_month = Moment(state.state_month).add(1, 'month');
    updateStatesMont(state);
  },
};
