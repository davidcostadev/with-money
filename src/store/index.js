import Vue from 'vue';
import Vuex from 'vuex';
import Moment from 'moment';
import mutations from './mutations';
import actions from './actions';

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    user: null,
    current_month: Moment(),
    state_month: Moment(),
    state_month_str: Moment().format('MMMM YY'),
    next_month_str: Moment().add(1, 'month').format('MMMM YY'),
    previous_month_str: Moment().subtract(1, 'month').format('MMMM YY'),
  },
  getters: {
    user(state) {
      return state.user;
    },
    current_month(state) {
      return state.current_month;
    },
    state_month(state) {
      return state.state_month;
    },
    state_month_str(state) {
      return state.state_month_str;
    },
    next_month_str(state) {
      return state.next_month_str;
    },
    previous_month_str(state) {
      return state.previous_month_str;
    },
  },
  mutations,
  actions,
});
