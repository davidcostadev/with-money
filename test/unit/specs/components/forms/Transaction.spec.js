import { createLocalVue, shallowMount } from '@vue/test-utils';
import VueMaterial from 'vue-material';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import Transaction from '../../../../../src/components/forms/Transaction';
import store from '../../../../../src/store';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueMaterial);
localVue.use(VueRouter);

const mock = {
  data: {
    data: [
      {
        id: 2,
        UserId: 1,
        name: 'Carteira',
        type: 'wallet',
        initalValue: '650.00',
        createdAt: '2018-08-04T14:33:21.000Z',
        updatedAt: '2018-08-04T14:33:21.000Z',
      },
      {
        id: 1,
        UserId: 1,
        name: 'Banco Inter',
        type: 'checking_account',
        initalValue: '505.18',
        createdAt: '2018-08-04T14:33:21.000Z',
        updatedAt: '2018-08-04T14:33:21.000Z',
      },
    ],
    pagination: {
      totalItems: 2,
      currentPage: 1,
      perPage: 100,
      totalPages: 1,
      nextPage: null,
      previousPage: null,
    },
  },
};

jest.mock('../../../../../src/services/accounts', () => ({
  get: () => Promise.resolve(mock),
}));

describe('Transaction Component', () => {
  describe('Render', () => {
    it('default', () => {
      const onSave = jest.fn();
      const wrapper = shallowMount(Transaction, {
        store,
        localVue,
        propsData: { onSave },
      });

      expect(wrapper.isVueInstance).toBeTruthy();
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
