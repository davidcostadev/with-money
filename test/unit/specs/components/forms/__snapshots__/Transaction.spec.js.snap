// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Component Render default 1`] = `
<form>
  <mdcard-stub>
    <mdcardcontent-stub>
      <mdfield-stub>
        <label
          for="name"
        >
          Name
        </label>
         
        <mdinput-stub />
      </mdfield-stub>
       
      <mdfield-stub>
        <input
          class="md-input"
          id="value"
          type="text"
        />
         
        <mdicon-stub>
          attach_money
        </mdicon-stub>
      </mdfield-stub>
       
      <mddatepicker-stub
        id="transactionDate"
      >
        <label>
          Transaction Day
        </label>
      </mddatepicker-stub>
       
      <mdfield-stub>
        <label
          for="type"
        >
          Type
        </label>
         
        <mdselect-stub>
          <mdoption-stub
            id="type-in"
          >
            In
          </mdoption-stub>
           
          <mdoption-stub
            id="type-out"
          >
            Out
          </mdoption-stub>
        </mdselect-stub>
      </mdfield-stub>
       
      <mdfield-stub>
        <label
          for="AccountId"
        >
          Account
        </label>
         
        <mdselect-stub />
      </mdfield-stub>
       
      <div>
        <mdswitch-stub
          class="md-primary"
        >
          IsPaid?
        </mdswitch-stub>
      </div>
    </mdcardcontent-stub>
     
    <mdcardactions-stub>
      <mdbutton-stub
        class="md-primary"
        id="save"
      >
        Save
      </mdbutton-stub>
    </mdcardactions-stub>
  </mdcard-stub>
</form>
`;
